runtime/compiler/env/J9SharedCache.cpp

```diff
@@ -810,10 +810,21 @@ UDATA *
 TR_J9JITaaSServerSharedCache::rememberClass(J9Class *clazz, bool create)
    {
    TR_ASSERT(_stream, "stream must be initialized by now");
-   // TODO: might want to keep a per-client cache of (class -> class chain)
-   // that were remembered to reduce the number of remote calls
+   PersistentUnorderedMap<J9Class *, UDATA *> & cache = TR::compInfoPT->getClientData()->getClassClainDataCache();
+      {
+      OMR::CriticalSection classChainDataMapMonitor(TR::compInfoPT->getClientData()->getClassChainDataMapMonitor());
+      auto it = cache.find(clazz);
+      if (it != cache.end())
+         return it->second;
+      }
    _stream->write(JITaaS::J9ServerMessageType::SharedCache_rememberClass, clazz, create);
-   return std::get<0>(_stream->read<UDATA *>());
+   UDATA * chainData = std::get<0>(_stream->read<UDATA *>());
+   if (chainData)
+      {
+      OMR::CriticalSection classChainDataMapMonitor(TR::compInfoPT->getClientData()->getClassChainDataMapMonitor());
+      cache.insert(std::make_pair(clazz, chainData));
```

Comment:
Looking at the implementation, if create is false then rememberClass() returns 1. It would be wrong to cache that value.

Also, the original rememberClass() has some diagnostic messages. Maybe we should add something similar to this implementation.